api/envoy/service/discovery/v3/discovery.proto

```diff
@@ -18,9 +19,19 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 
 // [#protodoc-title: Common discovery API components]
 
+// A specification of a resource used when subscribing or unsubscribing.
+message ResourceLocator {
+  // The resource name to subscribe to.
+  string name = 1;
+
+  // A set of constraints used to match against the dynamic parameters on the resource. This
+  // allows clients to select between multiple variants of the same resource.
+  DynamicParameterConstraints dynamic_parameter_constraints = 2;
```

Comment:
I realize it's out of scope for this PR, but do we anticipate in the stream-level optimization that we'll just have one of the maps at stream level and it will be unioned with the per-resource map? CC @adisuissa

Reply:
Yes, I think it makes sense to have a hierarchy of context-parameters (node -> stream -> resource_name), where each one can add and/or override the previous levels.

Reply:
I suspect the hierarchy we'll want is actually stream -> resource type -> resource name.  But we can figure out the details as part of Adi's design for version negotiation.